varname/ignore.py

```diff
@@ -156,87 +168,112 @@ def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
     def __repr__(self):
         return f'<IgnoreDecorated({self.ignore[0].__name__}, {self.ignore[1]})>'
 
-class IgnoreQualname(IgnoreElem):
-    """Ignore a function by qualified name"""
-    def __init__(self,
-                 ignore: Tuple[Optional[Union[ModuleType, str]], str]) -> None:
+class IgnoreModuleQualname(IgnoreElem):
+    """Ignore calls by qualified name in the module"""
+    def __init__(self, ignore: Tuple[ModuleType, str]) -> None:
         super().__init__(ignore)
-        self.module_flag = (
-            None if not ignore[0]
-            else 'module' if isinstance(ignore[0], ModuleType)
-            else 'filename'
-        )
 
         # check uniqueness of qualname
         self.qname_checked = False
-        if self.module_flag == 'module':
-            modfile = getattr(self.ignore[0], '__file__', None)
-            if modfile and path.isfile(modfile):
-                source = Source.for_filename(modfile, self.ignore[0].__dict__)
+        modfile = getattr(self.ignore[0], '__file__', None)
+        if modfile and path.isfile(modfile):
+            source = Source.for_filename(modfile, self.ignore[0].__dict__)
+            nobj = list(source._qualnames.values()).count(self.ignore[1])
+            # don't check for non-existence, as it could contain wildcards
+            assert nobj <= 1, (
+                f"Qualname {self.ignore[1]!r} in "
+                f"{self.ignore[0].__name__!r} refers to multiple objects."
+            )
+            self.qname_checked = True
+
+    def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
+        frame = frameinfos[frame_no].frame
+
+        module = inspect.getmodule(frame)
+        # Return earlier to avoid qualname uniqueness check
+        if module and module != self.ignore[0]:
+            return False
+
+        if not module and (
+                getattr(self.ignore, MODULE_IGNORE_ID_NAME, None) !=
+                frame.f_globals.get(MODULE_IGNORE_ID_NAME, '')
+        ):
+            return False
+
+        if not self.qname_checked: # // todo: coverage
+            source = Source.for_frame(frame)
+            if source.tree: # otherwise, no way to check
                 nobj = list(source._qualnames.values()).count(self.ignore[1])
-                assert nobj == 1, (
+                # don't check for non-existence, as it could contain wildcards
+                assert nobj <= 1, (
                     f"Qualname {self.ignore[1]!r} in "
-                    f"{self.ignore[0].__name__!r} "
-                    "doesn't exist or refers to multiple objects."
+                    f"{self.ignore[0].__name__!r} refers to multiple objects."
                 )
```

Comment:
Extract to a common function. Then you can probably use lru_cache instead of qname_checked